Add prompt input dialog Lua api

Delwing opened the pull request:
<!-- Keep the title short & concise so your mom can understand it -->
#### Brief overview of PR changes/additions

Adds non blocking prompt (input dialog), input as text or optionally as select box list.

![image](https://user-images.githubusercontent.com/3740628/102199579-14e50e80-3ec4-11eb-93e7-96b049a3f9c0.png)
![image](https://user-images.githubusercontent.com/3740628/102199663-2a5a3880-3ec4-11eb-8d3c-1e4e4c15997a.png)


#### Motivation for adding to Mudlet

Especially people coming from zMUD/cMUD background are for sure missing that functionality. Easy way to ask user for some input.

#### Other info (issues closed, discussion etc)

Commit:
add prompt lua api

vadi2 commented:
This is super similar to command lines, check it out:

```lua
myCommandLine = myCommandLine or Geyser.CommandLine:new({
  name = "myCommandLine", 
  x = "-20%", y = -40, 
  width = "20%", height = 40,
  stylesheet = "border: 1px solid silver;"
})

myCommandLine:setAction(function(commandLineInput)
    print(commandLineInput)
    myCommandLine:hide()
  end)
```

Maybe you can build on top of this to provide a similar function?

Delwing commented:
Similar, but not the same at all :)
Command line doesn't get focus right away and there is no select box option.

vadi2 commented:
Can you show how this'll be used in practice?

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -17923,3 +17925,75 @@ int TLuaInterpreter::setMapRoomExitsColor(lua_State* L)
+    auto *dialog = new QInputDialog(mudlet::self(), Qt::Dialog);
+    dialog->setWindowTitle(lua_tostring(L, 1));
+    dialog->setLabelText(lua_tostring(L, 2));
+    dialog->setOkButtonText(tr("OK"));
```
Is this and the next necessary or do they get populated automagically by the selection of the Qt translation by the GUI language setup code?

Delwing replied:
I was pretty sure, that this will work for translation, but I guess I have to check :)

Delwing replied:
I guess we can extend and add table with options, sure there is more to `QInputDialog` then exposed here. Actually while creating example for @vadi2 I realized terrible mistake I've made. There is really good reason to handle cancel clicks as well.

Delwing replied:
Actually `tr` does not work on translations I guess :/ We can delete both lines and we'll get standard OK and Cancel

Delwing replied:
>  Would it be a better UI experience if we could provide more access to the `QInputDialog` class - I note that there is scope for asking for a freeform string (possibly with a predefined value, and various echo modes including a password one that only shows the characters as they are entered but otherwise masked), an integer (with a predefined value and min / max settings) ....
> 
> ... perhaps it would be a more usable API if it took various "named" parameters from a table of options with one of them determining how the rest are to be used. There are a lot of potentially useful features in this Qt class and it might be possible to start off with something simple and develop it over time! For instance it could be set as read-only and pop up a value for the user to _copy_ from for their own use...

Maybe instead of creating type prompt and handling different option sets would be better to create prompt family functions, so like `promptForNumber`, `promptForPassword` and here not to include dozens of parameters it could have option table specific for flavour. But this is going a bit deeper into details, password dialog actually has a chance to be least used api of all time :) And `promptForNumber` while useful, can be added independently.

Additionally read-only popup is not that bad thing actually, but again - could be different API function, since it does not require accept function callback etc. I guess QT has some clipboard access capabilities, so it could be made even easier.

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -17923,3 +17925,75 @@ int TLuaInterpreter::setMapRoomExitsColor(lua_State* L)
+        lua_rawgeti(L, LUA_REGISTRYINDEX, callback);
+        lua_pushstring(L, dialog->textValue().toUtf8());
+        int error = lua_pcall(L, 1, 0, 0);
+        if (error != 0) {
```
:neckbeard: the `!= 0` is unnecessary here - in C/C++ any integer that is NOT `0` is logically `true`...! {Lua is different but this isn't Lua, :laughing: }

Delwing replied:
Oh OK, actually, copied that one from somewhere else, but good to know :)

SlySven reviewed (commented):
:thought_balloon: Would it be a better UI experience if we could provide more access to the `QInputDialog`  class - I note that there is scope for asking for a freeform string (possibly with a predefined value, and various echo modes including a password one that only shows the characters as they are entered but otherwise masked), an integer (with a predefined value and min / max settings) ....

... perhaps it would be a more usable API if it took various "named" parameters from a table of options with one of them determining how the rest are to be used. There are a lot of potentially useful features in this Qt class and it might be possible to start off with something simple and develop it over time! For instance it could be set as read-only and pop up a value for the user to *copy* from for their own use...

vadi2 commented:
I'm not sure that going down this path is a good idea, especially in the UI-experience department - UX standards have moved on from the 2000's zmud/cmud-style dialogs.

Delwing commented:
![prompt demo](https://user-images.githubusercontent.com/3740628/102247366-2815cf80-3f00-11eb-82f1-f0b22ae007e6.gif)

Delwing commented:
Not all users are tech savvy. It would simplify some configuration options in scripts just if I could ask them to input or pick specific option.
This is very easy and simple API and I am sure that many script developers would appreciate it. I personally always missed that kind of option.

Commit:
merge fresh development branch

Commit:
minor code clean up
change callback so it is called with first parameter as boolean defining whether input was accepted (OK) or rejected (Cancel)

Commit:
try to revert lfc changes

Commit:
revert .gitignore change

vadi2 commented:
If we made Mudlet for the tech savvy users, Mudlet wouldn't be Mudlet! I agree that we should make it simpler, but pop-up wizards for data entry are not exactly 2020 experience and we don't have a mission to bring the zmud nostalgia back. I have a few ideas on how we can do this better -

1) pre-filling the command line with the text to input, one after another to answer the survey
2) showing clickable text with clickable links

I think we can do better to solve the use case

Delwing commented:
I kinda still think that prompt like that has its place even nowdays. Well it's not like someone's forcing usage of that api to those who think otherwise :)

And there is also problem with setting focus on other command lines, it requires user to click there.

wiploo commented:
Hi, I've read this discussion on PR and even if I agree with @vadi2 about this question, I think that a stylable and maybe positionable box should be an improvement.

I can script some questions and related answers using command line (plain text o link to simulate select box), but the user can't type mud command meanwhile. If I want to ask a user an information, but leave it able to interact with his game, that's the solution.

Just my two cents :-D

Kebap commented:
This probably comes down to taste a bit. 
I for one would not like a random script take away my focus from the main command line or where I work, to a popup or second command line spontaneously. At least your suggested popup does not block me clicking back and finishing where I was working before.

Then again, I understand the need for a quick and easy API like the suggested `prompt(title, text, command)` - I think we can agree on this part at least so far.
If that is so, we only need to agree how the UX of entering data should be. 

There are lots of ideas floating around, prompt for text, selecting from drop down menu, accept only numbers, show hidden passwords, show text without any inputs... I can even think of more widgets, the sky is really the limit here, and I wonder how we want to navigate that slope.

Vadis suggestion seem to use plain text only, and I would like to see more details, on how exactly input and selection are handled.
Whereas Delwing suggested a more graphical style. Not sure if this works for users without mouse, only keyboard as well?

On another hand still, we are planning to bring Mudlet to mobile at some point, where scrolling a very long drop-down actually seems easier to use than typing text, or echoing all options to main text window to select from (if that was even suggested here)

After all this simple PR has some not-so-simple implications which should be pondered..

Delwing commented:
Actually this can be completely controlled by keyboard. Not so sure about alternatives proposed.

vadi2 commented:
Need to think more on this to solve the original problem in a better way. If we add things that allow people to create a bad user experience, we'll be the ones responsible for it, and I suspect it's too easy to make bad UX with this option.

Delwing commented:
I mean... I still don't get that hesitation. It's not like people are forced to use that. I personally have point where I would need that one. It would be great to use simple API than build this from scratch in geyser that would act as a poor replacement of that (no autofocus, no select box, no good keyboard support).

vadi2 commented:
> It's not like people are forced to use that

That's from the point of view of the scripter. From the point of view of the player, yeah, they will be forced to use that if it'll be in the package.

Delwing commented:
Yeah, but it's like their choice to use scripts or not :) And for less advanced users sometimes it's really better to just prompt for really important config options, especially at first use of scripts.
Scripters are closer to users usually and know how to adjust scripts to their needs, what's wrong and what's bad for them :)

Delwing commented:
So still not convinced? I would really like to use it as a first step to make initial configuration of my scripts as easy as possible.

vadi2 commented:
Yeah, I don't think this makes things as easy as possible. With printCmdLine you can walk people through a setup in a pretty easy way - just ask question after question with an echo and printCmdLine a pre-filled configuration alias. This has worked well for hundreds of players all without needing pop-ups flashing all the time or dialogs that steal focus under you.

vadi2 commented:
> Yeah, but it's like their choice to use scripts or not 

That's not how you build a platform... you can't tell people "Mudlet is great!" when scripts for Mudlet throw annoying popups at you.

> Scripters are closer to users usually and know how to adjust scripts to their needs, what's wrong and what's bad for them :)

Yes some, but not all, and I've seen some truly horrible code and designs in my time, giving people more easy to use tools for horrible design isn't the way to go. Better to give them easy tools for good design.

Delwing commented:
But printCmdLine actually is more intrusive than popup :)
Mapper itself uses kind of popups.

vadi2 commented:
It's not at all, it's not another window popping up and stealing focus on you. You start the setup and the command line gets prefilled with all the text you need to type. Very nice. You don't get ten popups one after another attacking you.

Kebap commented:
> With printCmdLine you can walk people through a setup in a pretty easy way - just ask question after question with an echo and printCmdLine a pre-filled configuration alias. This has worked well for hundreds of players all without needing pop-ups flashing all the time or dialogs that steal focus under you.

@vadi2 Can you show gif or video of this in action? I find it difficult to imagine. Thank you!

Edru2 commented:
Sorry for reviving this dead PR but I just discovered it :)

Very similar things should already be possible with a Geyser.CommandLine as Vadi suggested and a select box could be made with nested labels.
But as you already mentioned getting focus or using keyboard with it won't work.

I ever thought about adding something like setCmdLineFocus but I was hesitant because I had similar concerns as Kebap and Vadi. 

I may open an issue so we can discuss about setCmdLineFocus.
